tachyon/zk/r1cs/groth16/prove.h

```diff
@@ -46,11 +52,39 @@ Proof<Curve> CreateProofWithAssignment(const ProvingKey<Curve>& pk, const F& r,
                                        absl::Span<const F> full_assignments) {
   using G1AffinePoint = typename Curve::G1Curve::AffinePoint;
   using G2AffinePoint = typename Curve::G2Curve::AffinePoint;
+
+#if TACHYON_CUDA
+  using G1Bucket = typename math::VariableBaseMSMGpu<G1AffinePoint>::Bucket;
+  using G2Bucket = typename math::VariableBaseMSMGpu<G2AffinePoint>::Bucket;
+
+  gpuMemPoolProps props = {gpuMemAllocationTypePinned,
+                           gpuMemHandleTypeNone,
+                           {gpuMemLocationTypeDevice, 0}};
+  device::gpu::ScopedMemPool mem_pool = device::gpu::CreateMemPool(&props);
+
+  uint64_t mem_pool_threshold = std::numeric_limits<uint64_t>::max();
+  gpuError_t error = gpuMemPoolSetAttribute(
+      mem_pool.get(), gpuMemPoolAttrReleaseThreshold, &mem_pool_threshold);
+  CHECK_EQ(error, gpuSuccess);
+  device::gpu::ScopedStream stream = device::gpu::CreateStream();
+
+  math::VariableBaseMSMGpu<G1AffinePoint> msm(mem_pool.get(), stream.get());
+
+  device::gpu::ScopedMemPool mem_pool2 = device::gpu::CreateMemPool(&props);
+
+  error = gpuMemPoolSetAttribute(
+      mem_pool2.get(), gpuMemPoolAttrReleaseThreshold, &mem_pool_threshold);
+  CHECK_EQ(error, gpuSuccess);
+  device::gpu::ScopedStream stream2 = device::gpu::CreateStream();
```

Comment:
Can we reuse `mem_pool` and `stream` as well?

Reply:
Maybe for `mem_pool`, but for `stream`, I think it's better to separate as many as possible. See https://developer.download.nvidia.com/CUDA/training/StreamsAndConcurrencyWebinar.pdf I want to listen to @GideokKim 's opinion.

Reply:
### It is best to set up multiple streams.

If two kernel functions exist in the same stream, the kernel functions are necessarily executed sequentially. This means that the second kernel function will not be executed unless the execution of the previous kernel function is completed.

### In the case of memory pools, it is better not to be identical.

When using the same memory pool in different streams, record events using `cudaEventRecord(event, stream);` for one stream and `cudaStreamWaitEvent(streamB, event, 0);` for the other stream. It is recommended to use to wait for an event. I think it's better to let the CUDA driver do the optimization on its own. The CUDA driver can reuse memory opportunistically in the absence of explicit dependencies specified by the application.